core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/LogicalIndexExtractor.java

```diff
@@ -0,0 +1,394 @@
+
+    private static class Visitor
+            extends ExpressionRewriter<LogicalIndexContext>
+    {
+        private final Map<IrLabel, Set<IrLabel>> subsets;
+        private final ImmutableList.Builder<Symbol> layout;
+        private final ImmutableList.Builder<ValuePointer> valuePointers;
+        private final ImmutableSet.Builder<Symbol> classifierSymbols;
+        private final ImmutableSet.Builder<Symbol> matchNumberSymbols;
+        private final SymbolAllocator symbolAllocator;
+        private int copyIndex;
+
+        public Visitor(
+                Map<IrLabel, Set<IrLabel>> subsets,
+                ImmutableList.Builder<Symbol> layout,
+                ImmutableList.Builder<ValuePointer> valuePointers,
+                ImmutableSet.Builder<Symbol> classifierSymbols,
+                ImmutableSet.Builder<Symbol> matchNumberSymbols,
+                SymbolAllocator symbolAllocator)
+        {
+            this.subsets = requireNonNull(subsets, "subsets is null");
+            this.layout = requireNonNull(layout, "layout is null");
+            this.valuePointers = requireNonNull(valuePointers, "valuePointers is null");
+            this.classifierSymbols = requireNonNull(classifierSymbols, "classifierSymbols is null");
+            this.matchNumberSymbols = requireNonNull(matchNumberSymbols, "matchNumberSymbols is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+            this.copyIndex = 0;
+        }
+
+        @Override
+        protected Expression rewriteExpression(Expression node, LogicalIndexContext context, ExpressionTreeRewriter<LogicalIndexContext> treeRewriter)
+        {
+            return treeRewriter.defaultRewrite(node, context);
+        }
+
+        @Override
+        public Expression rewriteLabelDereference(LabelDereference node, LogicalIndexContext context, ExpressionTreeRewriter<LogicalIndexContext> treeRewriter)
+        {
+            Symbol reallocated = symbolAllocator.newSymbol(Symbol.from(node.getReference()), "_copy_" + copyIndex++);
+            layout.add(reallocated);
+            Set<IrLabel> labels = subsets.get(irLabel(node.getLabel()));
+            if (labels == null) {
+                labels = ImmutableSet.of(irLabel(node.getLabel()));
+            }
+            valuePointers.add(new ValuePointer(context.withLabels(labels).toLogicalIndexPointer(), Symbol.from(node.getReference())));
+            return reallocated.toSymbolReference();
+        }
+
+        @Override
+        public Expression rewriteSymbolReference(SymbolReference node, LogicalIndexContext context, ExpressionTreeRewriter<LogicalIndexContext> treeRewriter)
+        {
+            // symbol reference with no label prefix is implicitly prefixed with a universal row pattern variable (matches every label)
+            // it is encoded as empty label set
+            Symbol reallocated = symbolAllocator.newSymbol(Symbol.from(node), "_copy_" + copyIndex++);
```

Comment:
The `"_copy_" + copyIndex++` part is unnecessary. That's handled automatically by `SymbolAllocator`